cloudevents/extensions/opcua.md

```diff
@@ -0,0 +1,338 @@
+protocols, it therefore provides a recommendation to map known REQUIRED and
+OPTIONAL attributes using other extensions as well as defines its own extension
+attributes.
+
+## Notational Conventions
+
+As with the main [CloudEvents specification](../spec.md), the key words "MUST",
+"MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
+"RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as
+described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+However, the scope of these key words is limited to when this extension is used.
+For example, an attribute being marked as "REQUIRED" does not mean it needs to
+be in all CloudEvents, rather it needs to be included only when this extension
+is being used.
+
+## Mapping of REQUIRED Attributes
+
+### id
+
+MUST map to [Network Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.3#Table163)
+field `MessageId`.
+
+### source
+
+MUST either map to [Application
+Description](https://reference.opcfoundation.org/Core/Part4/v104/docs/7.1) field
+`applicationUri` of the OPC UA server or to a customer configured identifier
+like a unified namespace path.
+
+### type
+
+MUST map to [Data Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `MessageType`.
+
+## Mapping of OPTIONAL Attributes
+
+### datacontenttype
+
+MUST be `application/json` for OPC UA PubSub JSON payload and MAY be appended with
+`+gzip` when the payload is gzip compressed.
+
+### dataschema
+
+OPC UA provides type information as part of PubSub metadata messages, for non
+OPC UA consumers or when different payload encoding like Avro is used, it is
+REQUIRED to provide schema information (based on metadata information) in a
+separate format like [JSON schema](https://json-schema.org/specification) or
+[Avro schema](https://avro.apache.org/docs/1.11.1/specification/) or others. For
+those cases the attribute references the schema and is used for versioning.
+
+### subject
+
+For metadata, event and data messages (type one of `ua-metadata`, `ua-keyframe`,
+`ua-deltaframe`, `ua-event`), `subject` MUST map to either [Data Set Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `DataSetWriterId` or `DataSetWriterName`.
+
+For event messages (type equals to `ua-event`) `subject` MUST be appended with 
+"/" and [Base Event Type](https://reference.opcfoundation.org/Core/Part5/v104/docs/6.4.2)
+field `EventId`. 
+
+### time
+
+MUST map to [Data Set Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `Timestamp`.
+
+## Mapping for other extensions
+
+The following well-known extensions attributes are used for data messages and
+event messages (type one of `ua-keyframe`, `ua-deltaframe`, `ua-event`).
+
+### sequence
+
+Attribute as defined by [sequence extensions](./sequence.md) MUST map to [Data
```

Comment:
Is this one required for all OPC UA CloudEventss or is it optional?

Reply:
> The following well-known extensions attributes are used for data messages and
event messages (type one of `ua-keyframe`, `ua-deltaframe`, `ua-event`).

All CE attributes defined by other extensions are REQUIRED for type equals `ua-keyframe`, `ua-deltaframe` and `ua-event` and SHALL NOT be used for type equals `ua-metadata`.

Reply:
changed according to your suggestion